pkg/kubelet/kubelet_pods.go

```diff
@@ -1099,46 +1117,170 @@ func (kl *Kubelet) HandlePodCleanups() error {
+	podTerminationMap map[string]string
+	// killPod is the func which invokes runtime to kill the pod
+	killPod func(pod *v1.Pod, runningPod *kubecontainer.Pod, status *kubecontainer.PodStatus, gracePeriodOverride *int64) error
+	// getMirrorPodByPod calls pod manager's GetMirrorPodByPod
+	getMirrorPodByPod func(pod *v1.Pod) (*v1.Pod, bool)
+}
+
+// NewPodKiller returns a functional PodKiller
+func NewPodKiller(kl *Kubelet) PodKiller {
+	podKiller := &podKillerWithSlice{
+		podsToKill:        make([]*kubecontainer.PodPair, 0),
+		podKillingLock:    &sync.Mutex{},
+		podTerminationMap: make(map[string]string),
+		killPod:           kl.killPod,
+		getMirrorPodByPod: kl.podManager.GetMirrorPodByPod,
+	}
+	podKiller.cond = sync.NewCond(podKiller.podKillingLock)
+	return podKiller
+}
+
+// IsPodPendingTerminationByUID checks whether the pod for the given uid is pending termination
+func (pk *podKillerWithSlice) IsPodPendingTerminationByUID(uid types.UID) bool {
+	pk.podKillingLock.Lock()
+	defer pk.podKillingLock.Unlock()
+	if _, ok := pk.podTerminationMap[string(uid)]; ok {
+		return true
+	}
+	for _, pair := range pk.podsToKill {
+		if pair.RunningPod.ID == uid {
+			return true
+		}
+	}
+	return false
+}
+
+// IsPodPendingTerminationByPodName checks whether the given pod is in grace period of termination
+func (pk *podKillerWithSlice) IsPodPendingTerminationByPodName(podFullname string) bool {
+	pk.podKillingLock.Lock()
+	defer pk.podKillingLock.Unlock()
+	for _, pair := range pk.podsToKill {
+		if kubecontainer.GetPodFullName(pair.APIPod) == podFullname {
+			return true
+		}
+	}
+	for _, name := range pk.podTerminationMap {
+		if name == podFullname {
+			return true
+		}
+	}
+	return false
+}
+
+func (pk *podKillerWithSlice) markPodTerminated(uid string) bool {
+	pk.podKillingLock.Lock()
+	_, ok := pk.podTerminationMap[uid]
+	if ok {
+		klog.V(4).Infof("marking pod termination %q", uid)
+		delete(pk.podTerminationMap, uid)
+	}
+	pk.podKillingLock.Unlock()
+	return ok
+}
+
+// KillPod sends pod killing request to the killer
+func (pk *podKillerWithSlice) KillPod(pair *kubecontainer.PodPair) {
+	if pair.APIPod == nil {
+		pair.APIPod = pair.RunningPod.ToAPIPod()
+	}
+	pk.podKillingLock.Lock()
+	pk.podsToKill = append(pk.podsToKill, pair)
+	pk.podKillingLock.Unlock()
+	pk.cond.Signal()
+}
+
+// isPodBeingKilled adds the pod UID to set of pod UIDs which are being killed
+// It returns true when the pod UID is already in the set upon entry to the func
+func (pk *podKillerWithSlice) isPodBeingKilled(uid types.UID, podFullname string) bool {
+	pk.podKillingLock.Lock()
+	_, exists := pk.podTerminationMap[string(uid)]
+	if !exists {
+		pk.podTerminationMap[string(uid)] = podFullname
+	}
+	pk.podKillingLock.Unlock()
+	return exists
+}
+
+// wait waits for the arrival of subsequent pod killing request
+func (pk *podKillerWithSlice) wait() {
+	pk.podKillingLock.Lock()
+	defer pk.podKillingLock.Unlock()
+	for len(pk.podsToKill) == 0 {
+		pk.cond.Wait()
```

Comment:
this `Wait()`s with the `podKillingLock` lock held

Reply:
This is intentional, otherwise golang would panic.
Here is the body of Wait():
```
func (c *Cond) Wait() {
	c.checker.check()
	t := runtime_notifyListAdd(&c.notify)
	c.L.Unlock()
```
Please see the Unlock call.

Reply:
ok i see